src/core_modules/capture-core/metadataRetrieval/orgUnitName/orgUnitName.js

```diff
@@ -0,0 +1,173 @@
+import type { OrgUnitNames } from './orgUnitName.types';
+import type { QuerySingleResource } from '../../utils/api';
+
+// Avoid exporting displayNameCache to keep it truly private.
+// As a consequence all functions using it must be in this file.
+const displayNameCache = {};
+const maxBatchSize = 50;
+
+const displayNamesQuery = {
+    organisationUnits: {
+        resource: 'organisationUnits',
+        params: ({ filter }) => ({
+            fields: 'id,displayName',
+            filter: `id:in:[${filter}]`,
+            pageSize: maxBatchSize,
+        }),
+    },
+};
+
+const createBatches = (orgUnitIds: Array<string>): Array<Array<string>> => {
+    const reducedOrgUnitIds = Array.from(orgUnitIds
+        .filter(id => id)
+        .reduce((acc, id) => {
+            if (!displayNameCache[id]) {
+                acc.add(id);
+            }
+            return acc;
+        }, new Set()));
+
+    const batches = [];
+    for (let i = 0; i < reducedOrgUnitIds.length; i += maxBatchSize) {
+        batches.push(reducedOrgUnitIds.slice(i, i + maxBatchSize));
+    }
+
+    return batches;
+};
+
+// Works best with memoized input arrays.
+export const useOrgUnitNames = (orgUnitIds: Array<string>): {
+    loading: boolean,
+    orgUnitNames: ?OrgUnitNames,
+    error: any,
+} => {
+    const [fetching, setFetching] = useState(false);
+    const [fetchNextBatch, setFetchNextBatch] = useState(false);
+    const [requestedArray, setRequestedArray] = useState();
+    const [currentBatches, setCurrentBatches] = useState([]);
+    const [completedBatches, setCompletedBatches] = useState(0);
+    const [error, setError] = useState();
+
+    const ready = !fetching && orgUnitIds === requestedArray;
+
+    const batches = useMemo(
+        () => createBatches(orgUnitIds),
+        [orgUnitIds],
+    );
+    const filter = useMemo(
+        () => (fetching ? currentBatches[completedBatches].join(',') : ''),
+        [fetching, currentBatches, completedBatches],
+    );
+    const result = useMemo(
+        () => (ready ? orgUnitIds.reduce((acc, id) => {
+            acc[id] = displayNameCache[id];
+            return acc;
+        }, {}) : null),
+        [ready, orgUnitIds],
+    );
+
+    const onComplete = useCallback(({ organisationUnits }) => {
+        for (const { id, displayName } of organisationUnits.organisationUnits) {
+            displayNameCache[id] = displayName;
+        }
+        const completeCount = completedBatches + 1;
+        setCompletedBatches(completeCount);
+        if (completeCount === currentBatches.length) {
+            setFetching(false);
+        } else {
+            setFetchNextBatch(true);
+        }
+    }, [completedBatches, setCompletedBatches, currentBatches, setFetching, setFetchNextBatch]);
+
+    const onError = useCallback(
+        (fetchError) => {
+            setFetching(false);
+            setError(fetchError);
+        },
+        [setFetching, setError],
+    );
+
+    const { refetch } = useDataQuery(
```

Comment:
I had a look at `react-query`, but came to the conclusion that it wasn't the right tool for the job; for one it caches by default based on all parameters in a request. In the case of working lists we want to fetch several names in a single request and cache each name separately. According to Eirik it should be possible to set it up `react-query` so that it caches this way, but it implies having to battle with the system. Secondly I got the impression that one of the main merits with using `react-query` is the built-in memory management system, with stale times and automatic garbage collection. In the case of translated orgunit names we agreed there was no reason to expect changes during a single session. Thinking about it now it could be used for saving a little memory though, but hardly a significant amount..

Reply:
Sounds good! Thanks for the explanation 😄